File: Dizzo-Studio-Frontend/app/types/catalog.ts
import type { DesignDocument } from '~/lib/design/document';
import { i18nT } from '~/lib/i18n';

// Mirrors the backend's app/schemas/catalog.py. Money and lengths arrive as
// decimal strings ("149000.00", "226.00") — never floats — and are sent back
// the same way, so nothing is lost to rounding.

// Admin content in Russian and English: `{ ru: { name: … }, en: { … } }`.
// Uzbek stays in the resource's own fields; an empty value is untranslated.
export type TranslationLang = 'ru' | 'en';
export type Translations<T> = Partial<Record<TranslationLang, Partial<T>>>;

export type CatalogMethod = 'uv' | 'engrave';
export type ShapeKind = 'cylinder' | 'plane' | 'disc' | 'model';
export const MATERIALS = [
  'ceramic_glossy', 'ceramic_matte', 'glass_clear', 'glass_frosted', 'fabric', 'paper', 'plastic', 'metal', 'wood',
] as const;
export type Material = typeof MATERIALS[number];

export interface CylinderDims { diameter_mm: string; height_mm: string; handle: boolean; handle_gap_mm: string }
export interface PlaneDims { width_mm: string; height_mm: string; sides: 1 | 2 }
export interface DiscDims { diameter_mm: string }
export type ShapeDims = CylinderDims | PlaneDims | DiscDims | Record<string, never>;

export interface CylinderAnchor { start_mm: string; top_mm: string }
export interface PlaneAnchor { side: 'front' | 'back'; x_mm: string; y_mm: string }
export interface DiscAnchor { x_mm: string; y_mm: string }
export type Vec3 = [number, number, number];
// Decal projector on a GLB, in model space; {} = not placed yet.
// wrap_radius_mm: the print wraps round a cylinder (a mug) instead of lying flat.
// round: only the circle inside the area is printed (a clock's dial), so coverage is measured on it.
// corner_radius_mm: rounded corners of a rectangular face (the editor and the clock numerals follow them).
// dial: a clock face — new designs start with its numerals and minute marks.
export interface ModelAnchor {
  point: Vec3; normal: Vec3; up: Vec3; depth_mm: string; max_angle_deg: number; wrap_radius_mm?: string | null; round?: boolean;
  corner_radius_mm?: string | null; dial?: boolean;
}
export type AreaAnchor = CylinderAnchor | PlaneAnchor | DiscAnchor | ModelAnchor | Record<string, never>;
export interface AreaCamera { position: Vec3; target: Vec3 }

export interface ModelTransform {
  up_axis: 'y' | 'z';
  yaw_deg: 0 | 90 | 180 | 270;
  scale_ref: { a: Vec3; b: Vec3; mm: string } | null;
}

// Measured by the 3D configurator, sent when a GLB shape is marked ready.
export interface AreaCheck { coverage: number; stretched_share: number }

export function isPlacedModelAnchor(anchor: AreaAnchor): anchor is ModelAnchor {
  return 'point' in anchor && 'normal' in anchor && 'up' in anchor;
}

export interface AreaMethod {
  method: CatalogMethod;
  zone_x_mm: string;
  zone_y_mm: string;
  zone_w_mm: string;
  zone_h_mm: string;
  max_width_mm: string | null;
  max_height_mm: string | null;
  // A laser strip this wide (as tall as the zone) that slides across it:
  // all the method's layers go in one strip. null: the whole zone.
  strip_width_mm: string | null;
  min_font_mm: string | null;
  colors_allowed: boolean;
  dpi: number;
}

export interface PrintArea {
  id: number;
  key: string;
  name: string;
  width_mm: string;
  height_mm: string;
  anchor: AreaAnchor;
  camera: AreaCamera | null;
  placement_note: string;
  sort_order: number;
  // Symmetric partner (left ↔ right sleeve): same size and methods, zones
  // mirrored when pair_mirror.
  pair_key: string | null;
  pair_mirror: boolean;
  methods: AreaMethod[];
  translations?: Translations<{ name: string; placement_note: string }>; // admin API only
}

export interface Shape {
  id: number;
  name: string;
  description?: string; // a short line under the name in pickers (absent on an older API)
  kind: ShapeKind;
  dims: ShapeDims;
  model_url: string | null;
  model_media_id?: string | null; // admin API only
  model_transform: ModelTransform | null;
  mm_per_unit: string | null;
  status: 'draft' | 'ready';
  locked: boolean;
  archived: boolean;
  replaces_id: number | null; // a revision: takes that shape's place once ready
  areas: PrintArea[];
  translations?: Translations<{ name: string; description: string }>; // admin API only
}

/** Where a catalog picture came from: an admin's file, or drawn from the
 * 3D shape. A regenerate throws away the drawn ones and keeps the photos. */
export type PictureSourceKind = 'manual' | 'render';

export interface CatalogImage { media_id: string; url: string; source?: PictureSourceKind }

export interface VariantColor {
  id: number;
  name: string;
  hex: string;
  surcharge: string;
  is_available: boolean;
  sort_order: number;
  archived: boolean;
  images: CatalogImage[];
  // The one clean, colour-tinted picture the variant cards show. It is not
  // part of `images` — the gallery and the cards never fight over an index.
  card_image?: CatalogImage | null;
  card_image_url?: string | null;
  card_media_id?: string | null;
  translations?: Translations<{ name: string }>; // admin API only
}

export interface SpecItem { label: string; value: string }

// A clothing size, in the order the customer sees it. A variant with an
// empty list has no sizes (a mug) and none is ever chosen.
//
// print_scale is how much of the print area this size may use: the shape's
// area is measured for the LARGEST size (40 × 40 cm on a t-shirt is the
// maximum), and a smaller garment prints a proportionally smaller picture,
// centred on the same anchor. "1.00" = the whole area.
export interface VariantSize { label: string; surcharge: string; is_available: boolean; print_scale: string }

/** What the lead starts a new garment from; edited freely afterwards, with
 * the print scale a garment of that size normally takes (the backend's
 * services/catalog.py DEFAULT_PRINT_SCALES and migration 0029). */
export const DEFAULT_SIZES = ['S', 'M', 'L', 'XL', 'XXL'] as const;
export const DEFAULT_PRINT_SCALES: Record<string, string> = {
  '3XL': '1.00', 'XXXL': '1.00', 'XXL': '0.95', '2XL': '0.95', 'XL': '0.90', 'L': '0.85', 'M': '0.80', 'S': '0.75',
};
export const defaultPrintScale = (label: string): string => DEFAULT_PRINT_SCALES[label.trim().toUpperCase()] ?? '1.00';

export interface Variant {
  id: number;
  shape_id: number;
  name: string;
  short_description: string;
  description: string;
  specs: SpecItem[];
  base_price: string;
  methods: CatalogMethod[];
  material: Material;
  white_underbase: boolean;
  is_available: boolean;
  sort_order: number;
  archived: boolean;
  sizes: VariantSize[];
  main_image?: CatalogImage | null;
  main_image_url?: string | null;
  main_image_media_id?: string | null;
  images: CatalogImage[];
  colors: VariantColor[];
  translations?: Translations<{
    name: string; short_description: string; description: string; specs: SpecItem[];
    /** Only word labels: [{ label: 'Katta', name: 'Большой' }]; codes (S, XL, 42) are left out. */
    sizes: Array<{ label: string; name: string }>;
  }>; // admin API only
}

// The storefront shelf a product sits on: a category's slug. The shelves
// (name, icon, order) are managed in the admin.
export type ProductCategory = string;
export interface PublicCategory { slug: string; name: string; icon_svg: string; image_url: string | null }
export interface AdminCategory extends PublicCategory {
  id: number;
  image_media_id: string | null;
  sort_order: number;
  is_active: boolean;
  product_count: number;
  translations?: Translations<{ name: string }>;
}
/** A shelf as the chips and the catalog show it. */
export interface ShelfOption { value: string; label: string; icon_svg: string; image_url: string | null }

export interface PriceTier { method: CatalogMethod; min_cm2: string; max_cm2: string | null; surcharge: string }

export interface AdminCatalogProduct {
  id: number;
  slug: string;
  name: string;
  description: string;
  cover: CatalogImage | null;
  is_featured: boolean;
  is_available: boolean;
  sort_order: number;
  category: ProductCategory;
  archived: boolean;
  images: CatalogImage[];
  shapes: Shape[];
  variants: Variant[];
  price_tiers: PriceTier[];
  sellable: boolean;
  issues: string[];
  translations?: Translations<{ name: string; description: string }>;
}

export interface AdminCatalogListItem {
  id: number;
  slug: string;
  name: string;
  cover_url: string | null;
  is_featured: boolean;
  is_available: boolean;
  archived: boolean;
  sort_order: number;
  category: ProductCategory;
  variant_count: number;
  shape_count: number;
  sellable: boolean;
  from_price: string | null; // the cheapest sellable type and colour
}

export interface PublicProductCard {
  slug: string;
  name: string;
  cover_url: string | null;
  from_price: string;
  is_featured: boolean;
  category: ProductCategory;
}

export interface PublicColor {
  id: number;
  name: string;
  hex: string;
  surcharge: string;
  /** The colour's gallery, at most five pictures. Every one of them is of
   * this colour: none is reserved for anything else. */
  images: string[];
  /** The clean, colour-tinted shot the variant/colour cards show. */
  card_image_url: string | null;
}

export interface PublicVariant {
  id: number;
  shape_id: number;
  name: string;
  short_description: string;
  description: string;
  specs: SpecItem[];
  base_price: string;
  methods: CatalogMethod[]; // what the customer can use: variant ∩ the shape's areas
  material: Material;
  white_underbase: boolean;
  /** The plain, colourless base shot of the type — the first picture of the
   * detail gallery, whichever colour is chosen. */
  main_image_url: string | null;
  /** @deprecated The old name of `main_image_url`; the same picture. */
  variant_main_image?: string | null;
  images: string[];
  colors: PublicColor[]; // only colours on sale
  sizes: VariantSize[]; // in order; an out-of-stock size is listed, greyed out
}

export interface PublicShape {
  id: number;
  kind: ShapeKind;
  dims: ShapeDims;
  model_url: string | null;
  model_transform: ModelTransform | null;
  mm_per_unit: string | null;
  areas: PrintArea[];
}

export interface PublicProductDetail {
  slug: string;
  name: string;
  /** Prose about the product itself, above the types. */
  description: string;
  cover_url: string | null;
  /** Pictures of the product that hold for every type and colour (how it is
   * packed, a size chart): they close the detail gallery. */
  images: string[];
  from_price: string;
  variants: PublicVariant[]; // only variants on sale
  shapes: PublicShape[];
}

export interface QuoteLine { method: CatalogMethod; area_cm2: string; tier_min_cm2: string; tier_max_cm2: string | null; surcharge: string }

export interface Quote {
  variant_id: number;
  color_id: number;
  base_price: string;
  color_surcharge: string;
  size: string | null;
  size_surcharge: string;
  // The chosen size's share of the print area, and the largest painted area
  // it can hold per method (cm²) — what the price was capped to. Empty on a
  // product without sizes and on the largest size.
  print_scale: string;
  print_area_cm2: Partial<Record<CatalogMethod, string>>;
  methods: QuoteLine[];
  unit_price: string;
  quantity: number;
  total: string;
}

// Gallery designs (templates, app/schemas/template.py): made by admins in
// the Studio, offered on the variants they were checked against and shown
// with their pictures in the "Galereya".
export interface PublicTemplate {
  id: number;
  name: string;
  category: string;
  preview_url: string;
  variant_ids: number[];
  document: DesignDocument;
}

export interface AdminTemplate extends PublicTemplate {
  product_id: number;
  product_name: string;
  product_slug: string;
  is_active: boolean;
  sort_order: number;
  images: CatalogImage[];
  in_gallery: boolean;
  color_id: number | null;
  updated_at: string;
  translations?: Translations<{ name: string }>;
}

// The label maps below are getters: every read is in the page's current
// language (common.json), so `METHOD_LABELS[m]` stays reactive in templates.
const label = (key: string) => i18nT(`common.catalog.${key}`);

export const METHOD_LABELS: Record<CatalogMethod, string> = {
  get uv() { return label('method.uv'); },
  get engrave() { return label('method.engrave'); },
};

// What a customer reads before choosing how the design is put on.
export const METHOD_INFO: Record<CatalogMethod, { icon: string; short: string; details: string }> = {
  uv: {
    icon: 'lucide:palette',
    get short() { return label('methodInfo.uv.short'); },
    get details() { return label('methodInfo.uv.details'); },
  },
  engrave: {
    icon: 'lucide:zap',
    get short() { return label('methodInfo.engrave.short'); },
    get details() { return label('methodInfo.engrave.details'); },
  },
};

export const KIND_LABELS: Record<ShapeKind, string> = {
  get cylinder() { return label('kind.cylinder'); },
  get plane() { return label('kind.plane'); },
  get disc() { return label('kind.disc'); },
  get model() { return label('kind.model'); },
};

export const MATERIAL_LABELS: Record<Material, string> = {
  get ceramic_glossy() { return label('material.ceramic_glossy'); },
  get ceramic_matte() { return label('material.ceramic_matte'); },
  get glass_clear() { return label('material.glass_clear'); },
  get glass_frosted() { return label('material.glass_frosted'); },
  get fabric() { return label('material.fabric'); },
  get paper() { return label('material.paper'); },
  get plastic() { return label('material.plastic'); },
  get metal() { return label('material.metal'); },
  get wood() { return label('material.wood'); },
};

export interface DesignAsset {
  id: number;
  media_id: string;
  name: string;
  category: string;
  type: 'icon' | 'sticker' | 'photo';
  is_active: boolean;
  sort_order: number;
  url: string;
  px_w?: number | null;
  px_h?: number | null;
  created_at: string;
  updated_at: string;
}

File: Dizzo-Studio-Backend/app/models/catalog.py
"""Dynamic catalog: everything the storefront and the Studio show comes from
these tables, managed only through the admin panel.

Product -> Shape (dimensions, 3D rendering, print areas) and
Product -> Variant (picks a Shape, carries the base price) -> Color.
Nothing here is ever hard-deleted once it can be referenced by a cart or an
order: rows are archived (archived_at) instead.

Lengths are millimetres, stored as Numeric so validation is exact.
"""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal

from sqlalchemy import (
    JSON,
    Boolean,
    CheckConstraint,
    Column,
    DateTime,
    ForeignKey,
    Integer,
    Numeric,
    String,
    Table,
    Text,
    UniqueConstraint,
    false,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base, TimestampMixin, TranslatableMixin
from app.models.media import Media

METHODS = ("uv", "engrave")
SHAPE_KINDS = ("cylinder", "plane", "disc", "model")
SHAPE_STATUSES = ("draft", "ready")
MATERIALS = (
    "ceramic_glossy",
    "ceramic_matte",
    "glass_clear",
    "glass_frosted",
    "fabric",
    "paper",
    "plastic",
    "metal",
    "wood",
)

MM = Numeric(8, 2)
MONEY = Numeric(12, 2)

# Where a gallery picture came from: a human upload, or a renderer. Only
# 'render' rows may be thrown away and rebuilt by a bulk regenerate.
IMAGE_SOURCES = ("manual", "render")


class Product(TimestampMixin, TranslatableMixin, Base):
    __tablename__ = "products"

    id: Mapped[int] = mapped_column(primary_key=True)
    slug: Mapped[str] = mapped_column(String(120), unique=True, index=True)
    name: Mapped[str] = mapped_column(String(150))
    description: Mapped[str] = mapped_column(Text, default="")  # rich text: services.rich_text
    cover_media_id: Mapped[str | None] = mapped_column(ForeignKey("media.id", ondelete="SET NULL"), nullable=True)
    is_featured: Mapped[bool] = mapped_column(Boolean, default=True)
    is_available: Mapped[bool] = mapped_column(Boolean, default=True)
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    # Storefront shelf: the slug of a ProductCategory.
    category: Mapped[str] = mapped_column(String(40), default="boshqa", server_default="boshqa")
    archived_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    # Eager ("selectin") everywhere: the catalog is small, always read as a
    # whole, and async sessions can't lazy-load while serializing.
    cover: Mapped[Media | None] = relationship(lazy="selectin")
    images: Mapped[list[CatalogImage]] = relationship(
        foreign_keys="CatalogImage.product_id", order_by="CatalogImage.sort_order", lazy="selectin",
        cascade="all, delete-orphan",
    )
    shapes: Mapped[list[Shape]] = relationship(back_populates="product", order_by="Shape.id", lazy="selectin")
    variants: Mapped[list[Variant]] = relationship(
        back_populates="product", order_by="(Variant.sort_order, Variant.id)", lazy="selectin"
    )
    price_tiers: Mapped[list[PriceTier]] = relationship(
        back_populates="product", cascade="all, delete-orphan", order_by="(PriceTier.method, PriceTier.min_cm2)",
        lazy="selectin",
    )
    templates: Mapped[list[DesignTemplate]] = relationship(
        "DesignTemplate",
        back_populates="product",
        order_by="(DesignTemplate.sort_order, DesignTemplate.id)",
        lazy="selectin",
        cascade="all, delete-orphan",
    )


class ProductCategory(TimestampMixin, TranslatableMixin, Base):
    """A storefront shelf ("Idish-tovoq", "Kiyim"…), managed in the admin:
    its name, an SVG icon or a picture, and its place in the list. Products
    point to it by slug."""

    __tablename__ = "product_categories"

    id: Mapped[int] = mapped_column(primary_key=True)
    slug: Mapped[str] = mapped_column(String(40), unique=True, index=True)
    name: Mapped[str] = mapped_column(String(80))
    icon_svg: Mapped[str] = mapped_column(Text, default="", server_default="")
    image_media_id: Mapped[str | None] = mapped_column(ForeignKey("media.id", ondelete="SET NULL"), nullable=True)
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)

    image: Mapped[Media | None] = relationship(lazy="selectin")


class Shape(TimestampMixin, TranslatableMixin, Base):
    """A physical body ("Krujka 330 ml"): its dimensions, how it is drawn in
    3D and where it can be printed. Variants pick a shape. Once a design
    uses it, it is locked; changes are made on a revision — a draft copy
    (`replaces_id`) that takes its place, variants and all, once ready."""

    __tablename__ = "shapes"
    __table_args__ = (
        CheckConstraint(f"kind IN {SHAPE_KINDS}", name="kind"),
        CheckConstraint(f"status IN {SHAPE_STATUSES}", name="status"),
    )

    id: Mapped[int] = mapped_column(primary_key=True)
    product_id: Mapped[int] = mapped_column(ForeignKey("products.id", ondelete="CASCADE"), index=True)
    name: Mapped[str] = mapped_column(String(150))
    # A short line under the name in the admin's shape pickers ("oq, 330 ml").
    description: Mapped[str] = mapped_column(String(200), default="", server_default="")
    kind: Mapped[str] = mapped_column(String(16))
    # Kind-specific, validated by app.schemas.catalog: cylinder
    # {diameter_mm, height_mm, handle, handle_gap_mm}, plane {width_mm,
    # height_mm, sides}, disc {diameter_mm}, model {} (size comes from the GLB).
    dims: Mapped[dict] = mapped_column(JSON, default=dict)
    model_media_id: Mapped[str | None] = mapped_column(ForeignKey("media.id", ondelete="RESTRICT"), nullable=True)
    model_transform: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    mm_per_unit: Mapped[Decimal | None] = mapped_column(Numeric(12, 6), nullable=True)
    status: Mapped[str] = mapped_column(String(16), default="draft")
    locked_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    archived_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    replaces_id: Mapped[int | None] = mapped_column(ForeignKey("shapes.id", ondelete="SET NULL"), nullable=True)

    product: Mapped[Product] = relationship(back_populates="shapes")
    model: Mapped[Media | None] = relationship(lazy="selectin")
    areas: Mapped[list[PrintArea]] = relationship(
        back_populates="shape", cascade="all, delete-orphan", order_by="(PrintArea.sort_order, PrintArea.id)",
        lazy="selectin",
    )


class PrintArea(TimestampMixin, TranslatableMixin, Base):
    """A printable region of a shape, with its real size. `key` is stable
    (front, back, wrap…): when a customer switches to a variant with a
    different shape, layers move to the area with the same key."""

    __tablename__ = "print_areas"
    __table_args__ = (UniqueConstraint("shape_id", "key"),)

    id: Mapped[int] = mapped_column(primary_key=True)
    shape_id: Mapped[int] = mapped_column(ForeignKey("shapes.id", ondelete="CASCADE"), index=True)
    key: Mapped[str] = mapped_column(String(32))
    name: Mapped[str] = mapped_column(String(100))
    width_mm: Mapped[Decimal] = mapped_column(MM)
    height_mm: Mapped[Decimal] = mapped_column(MM)
    # Where the area sits on the shape, kind-specific: cylinder {start_mm,
    # top_mm}, plane {side, x_mm, y_mm}, disc {x_mm, y_mm}, model {point,
    # normal, rotation_deg, depth_mm, max_angle_deg}.
    anchor: Mapped[dict] = mapped_column(JSON, default=dict)
    camera: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    placement_note: Mapped[str] = mapped_column(Text, default="")
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    # Symmetric partner on the same shape (left ↔ right sleeve), by key. Both
    # sides point at each other and always share size and methods; with
    # pair_mirror the partner's zones — and a design copied onto it — are
    # mirrored left-right (content stays readable).
    pair_key: Mapped[str | None] = mapped_column(String(32), nullable=True)
    pair_mirror: Mapped[bool] = mapped_column(Boolean, default=True)

    shape: Mapped[Shape] = relationship(back_populates="areas")
    methods: Mapped[list[AreaMethod]] = relationship(
        back_populates="area", cascade="all, delete-orphan", order_by="AreaMethod.method", lazy="selectin"
    )


class AreaMethod(TimestampMixin, Base):
    """A decoration method allowed in an area, with the zone (inside the
    area, in area coordinates) it may be applied to and its limits."""

    __tablename__ = "area_methods"
    __table_args__ = (
        UniqueConstraint("area_id", "method"),
        CheckConstraint(f"method IN {METHODS}", name="method"),
    )

    id: Mapped[int] = mapped_column(primary_key=True)
    area_id: Mapped[int] = mapped_column(ForeignKey("print_areas.id", ondelete="CASCADE"), index=True)
    method: Mapped[str] = mapped_column(String(16))
    zone_x_mm: Mapped[Decimal] = mapped_column(MM)
    zone_y_mm: Mapped[Decimal] = mapped_column(MM)
    zone_w_mm: Mapped[Decimal] = mapped_column(MM)
    zone_h_mm: Mapped[Decimal] = mapped_column(MM)
    max_width_mm: Mapped[Decimal | None] = mapped_column(MM, nullable=True)
    max_height_mm: Mapped[Decimal | None] = mapped_column(MM, nullable=True)
    # A laser reaches only so far round a curved body: all the method's
    # layers go into one strip this wide (as tall as the zone), which slides
    # anywhere across the zone. None: the whole zone.
    strip_width_mm: Mapped[Decimal | None] = mapped_column(MM, nullable=True)
    min_font_mm: Mapped[Decimal | None] = mapped_column(MM, nullable=True)
    colors_allowed: Mapped[bool] = mapped_column(Boolean, default=True)
    dpi: Mapped[int] = mapped_column(Integer, default=300)

    area: Mapped[PrintArea] = relationship(back_populates="methods")


class Variant(TimestampMixin, TranslatableMixin, Base):
    """What the customer picks ("Xameleon"): a shape, a base price, the
    decoration methods it supports and the 3D material."""

    __tablename__ = "variants"
    __table_args__ = (CheckConstraint(f"material IN {MATERIALS}", name="material"),)

    id: Mapped[int] = mapped_column(primary_key=True)
    product_id: Mapped[int] = mapped_column(ForeignKey("products.id", ondelete="CASCADE"), index=True)
    shape_id: Mapped[int] = mapped_column(ForeignKey("shapes.id", ondelete="RESTRICT"), index=True)
    name: Mapped[str] = mapped_column(String(150))
    short_description: Mapped[str] = mapped_column(String(300), default="")
    description: Mapped[str] = mapped_column(Text, default="")
    specs: Mapped[list] = mapped_column(JSON, default=list)  # [{"label": "Hajmi", "value": "330 ml"}]
    # Clothing sizes, in the order the customer sees them:
    # [{"label": "M", "surcharge": "0.00", "is_available": true}]. Empty on
    # everything that has no size (a mug), and then none is ever chosen.
    sizes: Mapped[list] = mapped_column(JSON, default=list, server_default="[]")
    base_price: Mapped[Decimal] = mapped_column(MONEY)
    methods: Mapped[list] = mapped_column(JSON, default=list)  # subset of METHODS
    material: Mapped[str] = mapped_column(String(32), default="ceramic_glossy")
    # Transparent/dark bodies need a white ink layer under UV colours.
    white_underbase: Mapped[bool] = mapped_column(Boolean, default=False)
    is_available: Mapped[bool] = mapped_column(Boolean, default=True)
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    archived_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    product: Mapped[Product] = relationship(back_populates="variants", lazy="selectin")
    shape: Mapped[Shape] = relationship(lazy="selectin")
    images: Mapped[list[CatalogImage]] = relationship(
        foreign_keys="CatalogImage.variant_id", order_by="CatalogImage.sort_order", lazy="selectin",
        cascade="all, delete-orphan",
    )
    colors: Mapped[list[VariantColor]] = relationship(
        back_populates="variant", order_by="(VariantColor.sort_order, VariantColor.id)", lazy="selectin"
    )


class VariantColor(TimestampMixin, TranslatableMixin, Base):
    __tablename__ = "variant_colors"

    id: Mapped[int] = mapped_column(primary_key=True)
    variant_id: Mapped[int] = mapped_column(ForeignKey("variants.id", ondelete="CASCADE"), index=True)
    name: Mapped[str] = mapped_column(String(100))
    hex: Mapped[str] = mapped_column(String(7))
    surcharge: Mapped[Decimal] = mapped_column(MONEY, default=0)
    is_available: Mapped[bool] = mapped_column(Boolean, default=True)
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    archived_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    variant: Mapped[Variant] = relationship(back_populates="colors")
    images: Mapped[list[CatalogImage]] = relationship(
        foreign_keys="CatalogImage.color_id", order_by="CatalogImage.sort_order", lazy="selectin",
        cascade="all, delete-orphan",
    )


class CatalogImage(TimestampMixin, Base):
    """An ordered gallery image of exactly one product, variant or colour."""

    __tablename__ = "catalog_images"
    __table_args__ = (
        CheckConstraint(
            "(CASE WHEN product_id IS NULL THEN 0 ELSE 1 END)"
            " + (CASE WHEN variant_id IS NULL THEN 0 ELSE 1 END)"
            " + (CASE WHEN color_id IS NULL THEN 0 ELSE 1 END) = 1",
            name="one_owner",
        ),
        CheckConstraint(f"source IN {IMAGE_SOURCES}", name="source"),
    )

    id: Mapped[int] = mapped_column(primary_key=True)
    product_id: Mapped[int | None] = mapped_column(ForeignKey("products.id", ondelete="CASCADE"), nullable=True, index=True)
    variant_id: Mapped[int | None] = mapped_column(ForeignKey("variants.id", ondelete="CASCADE"), nullable=True, index=True)
    color_id: Mapped[int | None] = mapped_column(
        ForeignKey("variant_colors.id", ondelete="CASCADE"), nullable=True, index=True
    )
    media_id: Mapped[str] = mapped_column(ForeignKey("media.id", ondelete="RESTRICT"))
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    # "manual" (somebody uploaded it) or "render" (a renderer made it): a
    # bulk regenerate replaces its own output and leaves photographs alone.
    source: Mapped[str] = mapped_column(String(16), default="manual", server_default="manual")

    media: Mapped[Media] = relationship(lazy="selectin")


class PriceTier(TimestampMixin, Base):
    """Surcharge by painted area for one decoration method of a product. The
    tiers of a (product, method) always cover [0, ∞) without gaps or
    overlaps — they are replaced as a whole and validated on save."""

    __tablename__ = "price_tiers"
    __table_args__ = (
        UniqueConstraint("product_id", "method", "min_cm2"),
        CheckConstraint(f"method IN {METHODS}", name="method"),
    )

    id: Mapped[int] = mapped_column(primary_key=True)
    product_id: Mapped[int] = mapped_column(ForeignKey("products.id", ondelete="CASCADE"), index=True)
    method: Mapped[str] = mapped_column(String(16))
    min_cm2: Mapped[Decimal] = mapped_column(Numeric(10, 2))
    max_cm2: Mapped[Decimal | None] = mapped_column(Numeric(10, 2), nullable=True)  # NULL = no upper bound
    surcharge: Mapped[Decimal] = mapped_column(MONEY)

    product: Mapped[Product] = relationship(back_populates="price_tiers")


template_variants = Table(
    "design_template_variants",
    Base.metadata,
    Column("template_id", ForeignKey("design_templates.id", ondelete="CASCADE"), primary_key=True),
    Column("variant_id", ForeignKey("variants.id", ondelete="CASCADE"), primary_key=True),
)


class DesignTemplate(TimestampMixin, TranslatableMixin, Base):
    """A ready design the admin made in the Studio, shown in the "Galereya":
    customers start from it and edit it. It is offered on the variants it was
    checked against, and its images are library media, so any design may use
    them. Its pictures (mockups) and `in_gallery` put it on the public
    gallery; `color_id` is the colour it is shown and opened in."""

    __tablename__ = "design_templates"

    id: Mapped[int] = mapped_column(primary_key=True)
    product_id: Mapped[int] = mapped_column(ForeignKey("products.id", ondelete="CASCADE"), index=True)
    name: Mapped[str] = mapped_column(String(120))
    category: Mapped[str] = mapped_column(String(60), default="")  # a filter in the Studio ("Tug'ilgan kun")
    document: Mapped[dict] = mapped_column(JSON)
    preview_media_id: Mapped[str] = mapped_column(ForeignKey("media.id", ondelete="RESTRICT"))
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    sort_order: Mapped[int] = mapped_column(Integer, default=0)

    in_gallery: Mapped[bool] = mapped_column(Boolean, default=False, server_default=false(), index=True)
    color_id: Mapped[int | None] = mapped_column(
        ForeignKey("variant_colors.id", ondelete="SET NULL"), nullable=True
    )

    preview: Mapped[Media] = relationship(lazy="selectin")
    product: Mapped[Product] = relationship(back_populates="templates", lazy="selectin")
    variants: Mapped[list[Variant]] = relationship(
        secondary=template_variants, order_by="(Variant.sort_order, Variant.id)", lazy="selectin"
    )
    images: Mapped[list[DesignTemplateImage]] = relationship(
        order_by="DesignTemplateImage.sort_order", cascade="all, delete-orphan", lazy="selectin"
    )


class DesignTemplateImage(Base):
    """One gallery picture of a template (a mockup), in order."""

    __tablename__ = "design_template_images"

    id: Mapped[int] = mapped_column(primary_key=True)
    template_id: Mapped[int] = mapped_column(ForeignKey("design_templates.id", ondelete="CASCADE"), index=True)
    media_id: Mapped[str] = mapped_column(ForeignKey("media.id", ondelete="RESTRICT"))
    sort_order: Mapped[int] = mapped_column(Integer, default=0)

    media: Mapped[Media] = relationship(lazy="selectin")
